internal/rootcoord/root_coord.go

```diff
@@ -1523,12 +1534,128 @@ func (c *Core) Import(ctx context.Context, req *milvuspb.ImportRequest) (*milvus
+	}
+	return importJobResp, nil
+}
+
+func (c *Core) completeImportAsync(taskID int64) {
+	// First check if the import task has turned persisted state. Returns an error status if not after retrying.
+	// This could take a few or tens of seconds.
+	log.Info("start checking task persisted state",
+		zap.Int64("task ID", taskID))
+	getImportResp, err := c.checkImportTaskPersisted(taskID)
+	if err != nil {
+		// Task has not reached `ImportPersisted` state for a very long time, return and
+		// leave the task in its previous state forever, until it expires and cleaned up
+		// in the background.
+		log.Error("task not persisted yet after wait limit",
+			zap.Int64("wait limit (seconds)", int64(CheckTaskPersistedWaitLimit.Seconds())),
+			zap.Int64("task ID", taskID),
+			zap.Any("current task state", getImportResp.GetState()))
+		return
+	}
+	log.Info("start checking segment index ready state",
+		zap.Int64("task ID", taskID))
+	// Check index status. Note that checkSegmentIndexReady returns SUCCESS if:
+	// (1) there's no index defined on the collection, or
+	// (2) all indexes have been successfully built
+	checkIndexStatus, err := c.checkSegmentIndexReady(c.ctx, taskID, getImportResp.GetCollectionId(), getImportResp.GetSegmentIds())
+	if err != nil {
+		log.Warn(fmt.Sprintf("failed to wait for all index build to complete %s, but continue anyway", err.Error()))
+		return
+	}
+	if checkIndexStatus.GetErrorCode() != commonpb.ErrorCode_Success {
+		log.Warn(fmt.Sprintf("failed to wait for all index build to complete %s, but continue anyway", checkIndexStatus.Reason))
+		return
+	}
+
+	// Update import task state to `ImportState_ImportCompleted`.
+	// Retry on errors.
+	log.Info("trying to set task state as ImportComplete",
+		zap.Int64("task ID", taskID))
+	err = retry.Do(c.ctx, func() error {
+		status, err := c.ReportImport(c.ctx, &rootcoordpb.ImportResult{
+			TaskId: taskID,
+			State:  commonpb.ImportState_ImportCompleted,
+		})
+		if err != nil {
+			return err
+		}
+		if status.GetErrorCode() != commonpb.ErrorCode_Success {
+			return errors.New(status.GetReason())
+		}
+		return nil
+	}, retry.Attempts(reportImportAttempts))
+	if err != nil {
+		log.Error("failed to report import, we are not able to update the import task state to `ImportState_ImportCompleted`",
+			zap.Int64("task ID", taskID),
+			zap.Error(err))
+		// Data should not be visible if we failed to update task state.
+		// Everything has to start all over again.
+		return
+	}
+	log.Info("start unsetting isImporting state of segments",
+		zap.Int64s("segment IDs", getImportResp.GetSegmentIds()))
+	// Remove the `isImport` states of these segments only when the import task reaches `ImportState_ImportCompleted` state.
+	status, err := c.dataCoord.UnsetIsImportingState(c.ctx, &datapb.UnsetIsImportingStateRequest{
+		SegmentIds: getImportResp.GetSegmentIds(),
+	})
+	if err != nil {
+		log.Error("failed to unset importing state of all segments (could be partial failure)",
```

Comment:
we don't need to mark the task to fail or something?

Reply:
We can let the loop retry, and task may eventually fail when it times out.